File: aoc2016/16.py
QUIZ_NUMBER = "16"

class Solution:
    def run(fileName = QUIZ_NUMBER + ".in"):
        solution = Solution(fileName)
        solution.solve1()
        solution.solve2()

    def test():
        Solution.run(QUIZ_NUMBER + ".ex.in")

    def __init__(self, fileName):
        self.ins = []
        self.fileName = fileName
        fp = open(fileName, 'r')
        self.ins = list(map(str, fp.read()))
        fp.close()
        print("===", self.fileName, "===")

    def solve1(self):
        print("--- Part One ---")
        # l = 20
        l = 272

        # generate data
        a = self.ins[:]
        while len(a) < l:
            b = list(map(str, a))
            b.reverse()
            for j in range(len(b)):
                if b[j] == '1':
                    b[j] = '0'
                else:
                    b[j] = '1'
            a = list(map(str, "".join(a) + '0' + "".join(b)))
        a = a[0:l]

        # get checksum
        checksum = a[:]
        while len(checksum) % 2 == 0:
            newChecksum = []
            for i in range(0, len(checksum), 2):
                if checksum[i] == checksum[i+1]:
                    newChecksum.append('1')
                else:
                    newChecksum.append('0')
            checksum = newChecksum[:]
        print(f"Correct checksum for disk of length {l}: {"".join(checksum)}")


    def solve2(self):
        print("--- Part Two ---")
        # l = 20
        l = 35651584

        # generate data
        a = [False] * l
        for i in range(len(self.ins)):
            a[i] = True if self.ins[i] == '1' else False
        curL = len(self.ins)
        while curL < l:
            # print(curL)
            a[curL] = False
            for li in range(0, curL):
                if curL+li+1 >= l:
                    break
                a[curL+li+1] = not a[curL-li-1]
            curL += curL+1

        # get checksum
        checksum = a[:]
        while len(checksum) % 2 == 0:
            newChecksum = []
            for i in range(0, len(checksum), 2):
                if checksum[i] == checksum[i+1]:
                    newChecksum.append(True)
                else:
                    newChecksum.append(False)
            checksum = newChecksum[:]
        print(f"Correct checksum for disk of length {l}: ", end="")
        for b in checksum:
            if b:
                print('1', end="")
            else:
                print('0', end="")
        print()

def main():
    # Solution.test()
    Solution.run()

if __name__ == "__main__":
    main()

File: aoc2016/16.cpp
#include <bitset>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    // constexpr const int l = 20;
    // bitset<l> a;
    // a[l-0-1] = 1;
    // int curL = 5;

    // constexpr const int l = 272;
    constexpr const int l = 35651584;
    bitset<l> a;
    a[l-1-0] = 1;
    a[l-1-1] = 1;
    a[l-1-2] = 1;
    a[l-1-3] = 0;
    a[l-1-4] = 0;
    a[l-1-5] = 0;
    a[l-1-6] = 1;
    a[l-1-7] = 0;
    a[l-1-8] = 1;
    a[l-1-9] = 1;
    a[l-1-10] = 1;
    a[l-1-11] = 1;
    a[l-1-12] = 1;
    a[l-1-13] = 0;
    a[l-1-14] = 1;
    a[l-1-15] = 0;
    a[l-1-16] = 0;
    int curL = 17;

    while (curL < l) {
        a[l-1-curL] = 0;
        for (int li = 0; li < curL; li++) {
            if (l-1-(curL+li+1) < 0)
                break;
            a[l-1-(curL+li+1)] = ~a[l-1-(curL-li-1)];
        }
        curL += curL+1;
    }

    int n = l;
    while (n % 2 == 0) {
        for (int i = 0; i < n/2; i++) {
            a[i] = a[2*i] == a[2*i+1];
        }
        n = (n/2) + (n%2);
    }

    cout << "Correct checksum for disk of length " << l << ": ";
    for (int i = n-1; i >= 0; i--) {
        cout << a[i];
    }
    cout << endl;

    return 0;
}